### DISCUSSION TITLE
Should we update Java version?

### DISCUSSION THREAD

**Discussion by asvitkine** [triplea-game/triplea#10694] [General]
We're currently on Java 11. Should we update?

I think one downside of this version is on Mac, we don't have an M1 native version, so it runs in emulation (and thus much slower).

If we want to update, is there something more to do than changing `JavaVersion.VERSION_11` in build.gradle? When I tried setting something more recent, I get a build error...

**Comment by RoiEXLab** [comment]
> I think one downside of this version is on Mac, we don't have an M1 native version, so it runs in emulation (and thus much slower).

I'm pretty sure there are JDK 11 versions for M1. I checked the adoptium page, and shouldn't "aarch64" be the correct package for M1? https://adoptium.net/de/temurin/releases?version=11


> If we want to update, is there something more to do than changing `JavaVersion.VERSION_11` in build.gradle? When I tried setting something more recent, I get a build error...

I believe you're mixing multiple things here. One thing is the installed JDK version on your system. If this installed JDK version is older than the version you're trying to set this will not work. If you install a more recent JDK version this variable ensures the code is compiled according to the rules of Java 11 (See this [stackoverflow post](https://stackoverflow.com/a/10663480) for the difference of source and target version). So chances are you can install a JDK 17 and compile the code just fine.
**_However_** just because you updated your JDK and the version in the build file doesn't mean it will just work out of the box. There might be incompatible changes to syntax or to the standard library, so sometimes some minor changes are required to make it work. Also you need to consider that gradle might not support the newer JDK versions yet. AFAIK Gradle 8.0 (which is not out yet) is supposed to be the first release supporting the JDK 17. This JDK version is particularly interesting because it is supposed to be a "LTS" version for oracle (not for the OpenJDK but still). Also some libraries might not be compatible for some weird reason with the newer versions

Another thing that needs to be considered is the installer. Currently we bundle a JDK 11 with the installer, that of course would also need to be updated accordingly.

When it comes to language features I would be in favour of upgrading. Stuff like "records" is nice for tuple like structures without requiring lombok for example.

**Reply by asvitkine** [reply]
Oh, I guess I figured the Java version would be managed like any other depedency in gradle (where gradle would download it), but I guess it's a special?

As for Java 11 supporting arm64, that's probably on Linux. Java 11 was released in 2018, whereas M1 was announced in 2020, so it's not really possible that it could support arm64 on Mac (without major updates on a point release of Java 11, which seems unlikely).

**Reply by asvitkine** [reply]
For M1, maybe openjdk proper is only getting support for it in 17?

https://bugs.openjdk.org/browse/JDK-8272213

There are m1 ports of openjdk by azul and MS though: https://stackoverflow.com/questions/64788005/java-jdk-for-the-apple-m1-chip - but I guess we should just wait for gradle 8 and openjdk 17.

**Reply by RoiEXLab** [reply]
Gradle itself runs on java, so if no java exists on a system it couldn't start itself to manage it. It's a bit of a chicken-egg problem

**Reply by RoiEXLab** [reply]
And for dependencies gradle relies on maven central mostly. Maven central hosts java artifacts but not the runtime itself so it kind of makes sense

**Reply by RoiEXLab** [reply]
Oh and I just noticed the link i provided was not specific enough.
Here you can see the adoptium JDK 11 binaries for all architectures including M1 (aarch64) https://github.com/adoptium/temurin11-binaries/releases/tag/jdk-11.0.15%2B10
You can choose between a pkg installer and a tar.gz to extract the files. I'm not sure how you installed java up until this point, but if you use a package manager like homebrew I'm 99% sure they have a way to install exactly this version of the JDK on your machine.

**Reply by asvitkine** [reply]
Interesting that there's a 11 version for mac arm, I don't understand how that works with the openjdk ticket saying it's a jdk 17 thing.

I understand gradle needs java, but I don't see why the java gradle runs for itself needs to be the same that's used to compile and run the given project. I mean, I understand that might be the decision they made, but seems not a great design decision by gradle.

**Reply by DanVanAtta** [reply]
Java Version comes into play in a few places:
- JDK syntax & build version that are specified in gradle (which version to compile the code to, and which language features to recognize)
- various build tools & dependencies. Notably 'error prone' has been finicky with newer java versions and we had a lot of trouble getting onto java 11 as it would only work with java 9
- install4j bundled JDKs

The very latter is something we can swap out for our own bundled JDKs. If there is a Mac JDK that we can bundle that is not run on an emulator, that is something we could readily swap out.

**Reply by RoiEXLab** [reply]
> If there is a Mac JDK that we can bundle that is not run on an emulator, that is something we could readily swap out.

There is but we have to distribute 2 mac installers then, we can't just swap it out.

**Reply by DanVanAtta** [reply]
Why can't we just swap it out? We used to bundle our own, the config looks similar, the option is "use a pre-created JRE bundle":
![Screenshot from 2022-06-19 14-24-45](https://user-images.githubusercontent.com/12397753/174500891-e9fa70a9-4b08-41b5-bd66-8ab32d98f1b4.png)

**Reply by RoiEXLab** [reply]
@DanVanAtta I mean we can swap it out, but "older" devices running macOs are still using Intel processors (x86 architecture) and we probably want to keep supporting those. The newer ARM JDK will most likely not run on x86

**Comment by RoiEXLab** [comment]
Good question about the ticket. For the most part it should be possible to just change the compiler to generate different machine code, but maybe they backported some code changes?

Regarding the gradle java thingy: I agree that this seems like a poor design choice, but I honestly can't think of a language ecosystem that has its version managed as well on a project basis. For python things like virtual environments exist, but you kinda have to create them on your own for specific projects to my knowledge. C/C++ doesn't even have a proper widely adopted dependency management system and if your compiler doesn't support a certain feature yet you're out of luck. JavaScript also requires you to have a recent nodejs version/hope the the browser supports it.
So I think this problem is larger than the java ecosystem

**Comment by DanVanAtta** [comment]
Now that we are a year later.. JDK 17 is pretty attractive for its multi-line strings. I'd be happy to see a JDK version upgrade.
